design/healcheck-endpoint.md

```diff
@@ -0,0 +1,134 @@
+* Additional connections allocated in the thread pools.
+* Often requires integrating with additional subsystems (e.g. JCA for database checks).
+* Complicated to ensure that `timeout` parameter is properly respected.
+
+All the above pain points are not applicable for the Passive Mode. Essentially, all database, User Federation LDAP servers
+and External IdPs probes are free of charge as we report success/failure calls for them in the Metrics components.
+With this approach, the Healthcheck Endpoint becomes an alternative reporting tool for the Metrics.
+
+This design assumes implementing Health Endpoint in Passive Mode (with possible elements of active checks).
+
+#### Microprofile Health
+
+[Eclipse Microprofile Health](https://download.eclipse.org/microprofile/microprofile-health-3.0-RC4/microprofile-health-spec.html)
+project provides capabilities for reporting Liveness and Readiness probes. In the contrast to Microprofile Metrics,
+the Health project doesn't provide a centralized class to access all configured checks. The SmallRye implementation
+provides `SmallRyeHealthReporter` (see [the code](https://github.com/smallrye/smallrye-health/blob/73e9ea8c31498dab7753f4a428cb424bf9fbaa1a/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L49))
+, that is a CDI-annotated Bean that contains references to `HealthCheck` implementations. The `HealthCheck` interface
+implementations provides a specific check - e.g. check if a database is up.
+
+Keycloak will provide either its own implementation of a class similar to `SmallRyeHealthReporter`, or
+a CDI-annotated bean that will be used as a bridge between Microprofile Health, and the rest of the Keycloak code. The
+latter solution is preferred as the out-of-the-box support has been already implemented in Quarkus.
+
+Individual Healthcheck implementations, such as database check, External IdP check or LDAP check need to be provided
+as `HealthCheck` class implementations.
+
+#### Database check
+
+Agroal (the lightweight database connection pool) publishes a set of metrics into the main Quarkus Metrics Registry
+(see [the code](https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/metrics/AgroalMetricsRecorder.java#L37-L113)).
+It also provides an active healthcheck (see [the code](https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java)).
+
+The goal of the implementation is to provide a specialized check that is a hybrid between an active and a passive mode.
+When Keycloak is serving traffic, the implementation needs to track the database health by observing metrics reported
+by the Agroal framework. However, when Keycloak is idle, the implementation needs to either ask Agroal to maintain at least
+one, healthy connection or perform an active check.
+
+#### Healthcheck API for Liveness and Readiness probes
```

Comment:
It's not completely clear here, but I don't believe it is best practice for a readiness or liveness endpoint to check external dependencies. What is restarting keycloak pods going to do if the database is down? (Just make a situation worse, would be my guess)

Reply:
The Healthcheck Endpoint only checks the database at the moment (in it's readiness probe). Perhaps, in the future we will include a remote Infinispan cluster. But this needs a bit of experimentation. 

The Liveness check only checks if the Keycloak REST interface is up. If it's not - something terrible happened at the boot time. So there's a good argument to restart it.